prov/sm2/src/sm2_coordination.c

```diff
@@ -0,0 +1,574 @@
+}
+
+ssize_t sm2_mmap_cleanup(struct sm2_mmap *map)
+{
+	int err1, err2;
+
+	err1 = close(map->fd);
+	if (err1)
+		FI_WARN(&sm2_prov, FI_LOG_AV,
+			"Failed to close fd %d with error code %d", map->fd,
+			err1);
+
+	err2 = munmap(map->base, map->size);
+	if (err2)
+		FI_WARN(&sm2_prov, FI_LOG_AV,
+			"Failed unmap shared memory region with error code %d",
+			err2);
+
+	return err1 ? err1 : err2;
+}
+
+ssize_t sm2_file_open_or_create(struct sm2_mmap *map_shared)
+{
+	pthread_mutexattr_t att;
+	struct sm2_mmap map_ours;
+	char template[SM2_NAME_MAX];
+	struct sm2_coord_file_header *header, *tmp_header;
+	struct sm2_ep_allocation_entry *entries;
+	int fd, common_fd, err, tries, item;
+	bool have_file_lock = false;
+
+	sprintf(template, "%s/fi_sm2_pid%d_XXXXXX", SM2_COORDINATION_DIR,
+		getpid());
+
+	/* Assume we are the first process here.
+	 * Try to open a tmpfile file in the shm directory
+	 */
+	fd = mkostemp(template, O_RDWR);
+	assert(fd > 0);
+	err = ftruncate(fd, sizeof(*header));
+	if (OFI_UNLIKELY(err))
+		return -FI_ENOMEM;
+
+	header = sm2_mmap_map(fd, &map_ours);
+
+	pthread_mutexattr_init(&att);
+	pthread_mutexattr_setpshared(&att, PTHREAD_PROCESS_SHARED);
+	pthread_mutex_init(&header->write_lock, &att);
+
+	header->file_version = SM2_VERSION;
+	header->ep_region_size = sm2_calculate_size_offsets(NULL, NULL);
+	header->ep_allocation_offset = sizeof(*header);
+	header->ep_allocation_offset = NEXT_MULTIPLE_OF(
+		header->ep_allocation_offset, SM2_MAX_UNIVERSE_SIZE);
+	header->ep_regions_offset = header->ep_allocation_offset +
+				    (SM2_MAX_UNIVERSE_SIZE * sizeof(*entries));
+	header->ep_regions_offset = NEXT_MULTIPLE_OF(header->ep_regions_offset,
+						     SM2_MAX_UNIVERSE_SIZE);
+
+	/* allocate enough space in the file for all our allocations, but no
+	 * data exchange regions yet.
+	 */
+	sm2_mmap_remap(&map_ours, header->ep_regions_offset);
+	header = (struct sm2_coord_file_header *) map_ours.base;
+	entries = sm2_mmap_entries(&map_ours);
+	for (item = 0; item < SM2_MAX_UNIVERSE_SIZE; item++)
+		entries[item].pid = 0;
+
+	pthread_mutex_lock(&header->write_lock);
+
+	/* Make sure the header is written before we link the file,
+	 * flush file
+	 */
+	atomic_mb();
+
+	tries = SM2_STARTUP_MAX_TRIES;
+	do {
+		/* create a hardlink to our file with the common name.
+		 *  - on success: we hold the lock to a newly
+		 *		  created coordination file.
+		 *  - on failure: the file already exists, we should try opening
+		 *  		  it.
+		 */
+		err = link(template, SM2_COORDINATION_FILE);
+		if (0 == err) {
+			/* We are using the memory we initialized.  Duplicate it
+			 * and leave the map open. */
+			memcpy(map_shared, &map_ours, sizeof(map_ours));
+			have_file_lock = true;
+			break;
+		}
+
+		common_fd = open(SM2_COORDINATION_FILE, O_RDWR);
+		if (common_fd > 0) {
+			tmp_header = sm2_mmap_map(common_fd, map_shared);
+			assert(map_shared->size >= sizeof(*tmp_header));
+			err = pthread_mutex_trylock(&tmp_header->write_lock);
```

Comment:
How is the correct order of write_lock's determined between header->write_lock and tmp_header->write_lock?  I'm assuming these locks are shared between processes?

Reply:
If we get in here... that means that we failed to link our file to be the main file, b/c the main file already exists.  This means that we must try to get the write lock on the already existing file.  We poll it instead of blocking on it b/c we hope to add error handling for what we should do if the PID holding the lock is dead.